Fixes #229 - Sdp answer only contains actually configured codecs

shardik86 opened the pull request:
AudioChannel instance was created populating its supportedFormats variable with every possible codec. I modified that passing to ChannelManager only the iterator of actually configured codecs, so that AudioChannle Instance can be created initializing supportedFormats accordingly.

hrosa commented on bootstrap/src/main/java/org/mobicents/media/server/bootstrap/ioc/provider/ChannelsManagerProvider.java:
```diff
@@ -51,6 +51,7 @@ public ChannelsManager get() {
         ChannelsManager channelsManager = new ChannelsManager(this.udpManager);
         channelsManager.setScheduler(mediaScheduler);
         channelsManager.setJitterBufferSize(config.getMediaConfiguration().getJitterBufferSize());
+        channelsManager.setCodecs(config.getMediaConfiguration().getCodecs());
```
Change the getCodecs method to return an array instead of iterator.

hrosa commented on io/rtp/src/main/java/org/mobicents/media/server/impl/rtp/ChannelsManager.java:
```diff
@@ -57,6 +58,9 @@
     private int jitterBufferSize=50;
     
+    //Supported Formats --> It contains only codecs actually configured
+    private Iterator<String> codecs;
```
Use an array instead of iterator.

hrosa commented on io/rtp/src/main/java/org/mobicents/media/server/impl/rtp/ChannelsManager.java:
```diff
@@ -112,6 +116,14 @@ public void setJitterBufferSize(int jitterBufferSize) {
     	this.jitterBufferSize=jitterBufferSize;
     }        
     
+    public Iterator<String> getCodecs() {
```
Use array instead of iterator

hrosa commented on io/rtp/src/main/java/org/mobicents/media/server/impl/rtp/ChannelsManager.java:
```diff
@@ -112,6 +116,14 @@ public void setJitterBufferSize(int jitterBufferSize) {
+		return codecs;
+	}
+    
+    public void setCodecs(Iterator<String> iterator) {
```
Use array instead of iterator

hrosa commented on io/rtp/src/main/java/org/mobicents/media/server/impl/rtp/channels/AudioChannel.java:
```diff
@@ -39,7 +43,59 @@
+		//super.supportedFormats = AVProfile.audio;
+		this.supportedFormats = new RTPFormats();
+		Iterator<String> codecs = channelsManager.getCodecs();
+		while(codecs.hasNext()) {
```
Use AVProfile.getFormat(int) instead of creating the formats yourself.

hrosa commented on io/rtp/src/main/java/org/mobicents/media/server/impl/rtp/channels/AudioChannel.java:
```diff
@@ -39,7 +43,59 @@
 		super(MEDIA_TYPE, wallClock, channelsManager);
 //		super.supportedFormats = super.buildRTPMap(AVProfile.audio);
-		super.supportedFormats = AVProfile.audio;
+		//super.supportedFormats = AVProfile.audio;
```
Delete this line

hrosa commented:
Hello @shardik86,

I've reviewed your commit. Please check my comments.
Can you also provide a unit test to make sure we don't have regressions in the future?

Thank you for contributing! Much appreciated.

shardik86 commented:
Hello @hrosa,
I've seen your comments and modified my code. 
Since I am a beginner I don't know how to create a proper unit test, but I see that there are many tests regaring creation of media channels and formats negotiation. Since my code caused many existing tests to fail since everytime ChannelsManager is created the array of codecs must be supplied in order to create new audio channels, then I modified all those tests adding a custom array of codecs. Am I breaking any programming rule or best practice by doing that?

hrosa commented on core/src/main/java/org/mobicents/media/core/configuration/MediaConfiguration.java:
```diff
@@ -107,8 +106,8 @@ public void addCodec(String codec) {
-    public Iterator<String> getCodecs() {
-        return this.codecs.iterator();
+    public String[] getCodecs() {
+        return this.codecs.toArray(new String[]{});
```
specify array length (based on codecs.size) to optimize.
http://stackoverflow.com/a/5982478/4142153

hrosa commented:
Hi @shardik86 

If the requirements for building a channel manager changed, then you have to update the test cases to reflect those changes. It's only normal :)

hrosa commented on io/rtp/src/main/java/org/mobicents/media/server/impl/rtp/channels/AudioChannel.java:
```diff
@@ -43,45 +40,37 @@
 			case "gsm":
-				RTPFormat gsm = new RTPFormat(3, FormatFactory.createAudioFormat("gsm", 8000), 8000);
-				supportedFormats.add(gsm);
+				supportedFormats.add(AVProfile.getFormat(8));
```
GSM format is not 8! Please fix.

hrosa commented:
Hi @shardik86 

I noticed the GSM codec is not being properly loaded/negotiated. Please fix.

Thank you

hrosa commented:
Thank you for your efforts @shardik86 
You've been acknowledged https://github.com/Restcomm/mediaserver

If you wish to keep contributing to the project, feel free to browse our [list of community issues](https://github.com/RestComm/mediaserver/issues?q=is%3Aissue+is%3Aopen+label%3Ahelp-wanted)

We appreciate it!
